crates/notion-core/src/distro/package.rs

```diff
@@ -363,6 +370,88 @@ impl PackageVersion {
+    /// * the json config files
+    /// * the shims
+    /// * the unpacked and initialized package
+    ///
+    /// This optionally clears the inventory for this package, removing:
+    /// * the downloaded tarball and shasum files
+    pub fn uninstall(name: String, remove_all: bool) -> Fallible<()> {
+        // if the package config file exists, use that to remove any installed bins and shims
+        let package_config_file = path::user_package_config_file(&name)?;
+        if package_config_file.exists() {
+            let package_config = PackageConfig::from_file(&package_config_file)?;
+
+            for bin_name in package_config.bins {
+                PackageVersion::remove_config_and_shims(&bin_name, &name)?;
+            }
+
+            fs::remove_file(package_config_file).with_context(file_deletion_error)?;
+        } else {
+            // there is no package config - check for orphaned binaries
+            let user_bin_dir = path::user_bin_dir()?;
+            if user_bin_dir.exists() {
+                let orphaned_bins = binaries_from_package(&name)?;
+                for bin_name in orphaned_bins {
+                    PackageVersion::remove_config_and_shims(&bin_name, &name)?;
+                }
+            }
+        }
+
+        // if any unpacked and initialized packages exists, remove them
+        let package_image_dir = path::package_image_root_dir()?.join(&name);
+        if package_image_dir.exists() {
+            fs::remove_dir_all(package_image_dir).with_context(file_deletion_error)?;
+        }
+
+        // if the `--full` option was passed in, remove package files from the inventory
+        if remove_all {
+            // all cached files for the package name
+            let re = Regex::new(&format!(r"^{}-\d+\.\d+\.\d+.*\.(tgz|shasum)$", name)).unwrap();
+            let inventory_dir = path::package_inventory_dir()?;
+
+            let cached_files = files_matching(&inventory_dir, &re)?;
+            for file in cached_files {
+                fs::remove_file(&file).with_context(file_deletion_error)?;
+            }
+
+            println!("Removed cached files for '{}'", name);
+        }
+
+        println!("Package '{}' uninstalled", name);
+        Ok(())
+    }
+
+    fn remove_config_and_shims(bin_name: &str, name: &str) -> Fallible<()> {
+        let shim = path::shim_file(&bin_name)?;
+        fs::remove_file(shim).with_context(file_deletion_error)?;
+        let config_file = path::user_tool_bin_config(&bin_name)?;
+        fs::remove_file(config_file).with_context(file_deletion_error)?;
+        println!("Removed executable '{}' installed by '{}'", bin_name, name);
+        Ok(())
+    }
+}
+
+/// Reads the contents of a directory and returns a Vec containing the names of
+/// all the binaries installed by the input package.
+pub fn binaries_from_package(package: &str) -> Fallible<Vec<String>> {
+    let bin_config_dir = path::user_bin_dir()?;
+    Ok(read_dir_eager(&bin_config_dir)?
```

Comment:
This and the implementation of `files_matching` in `fs` are very similar. Would it be possible to abstract them in some way to prevent duplication of the boilerplate parts of it?

```
read_dir_eager(dir)?
    .filter(|(_, metadata)| metadata.is_file())
    .filter_map(|(entry, _)| {
    })
    .collect::<Vec<String>>())
```

Perhaps by making `files_matching` into a more general implementation that accepts a `FnOnce(PathBuf) -> Option<String>` and calls that, so we can implement each of these filters (config or regex) with only their specific logic?

Reply:
That's a good idea - I'll see if I can abstract that out

Reply:
I was able to refactor this into a `dir_entry_match` function that takes `FnMut(DirEntry) -> Option<T>`, so it can be used by both implementations.